### ISSUE TITLE
EF Core and SqlClient instrumentation enabled at the same time produce wrong spans

### ISSUE THREAD

**Issue by virtes** [open-telemetry/opentelemetry-dotnet#5009]
# Bug Report

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):

* OpenTelemetry - 1.7.0-alpha.1
* OpenTelemetry.Exporter.Console - 1.7.0-alpha.1
* OpenTelemetry.Exporter.OpenTelemetryProtocol - 1.7.0-alpha.1
* OpenTelemetry.Instrumentation.EntityFrameworkCore - 1.0.0-beta.8
* OpenTelemetry.Instrumentation.SqlClient - 1.6.0-beta.2

Runtime version (e.g. `net462`, `net48`, `netcoreapp3.1`, `net6.0` etc. You can
find this information from the `*.csproj` file):

* net6.0
* net7.0

## Symptom

Using `AddSqlClientInstrumentation` and `AddEntityFrameworkCoreInstrumentation` at the same time leads to miss parent span.

**What is the expected behavior?**

Exports two spans:
1. EF Core span
2. SQL Client span, with EF Core span as parent 

**What is the actual behavior?**
1. EF Core span **is missed** in the output
2. SQL Client span, with EF Core span as parent

Output from console exporter - I have only SQL Client span, which has ParentSpanId of EF Core span.

```
Activity.TraceId:            916f8741531547a078d9966039e4f905
Activity.SpanId:             96a036d8982f1155
Activity.TraceFlags:         Recorded
Activity.ParentSpanId:       8580d5d7086bf8c8
Activity.ActivitySourceName: OpenTelemetry.Instrumentation.SqlClient
Activity.DisplayName:        lsnext_domain
Activity.Kind:               Client
Activity.StartTime:          2023-10-31T13:33:26.4747510Z
Activity.Duration:           00:00:13.8057920
Activity.Tags:
    db.system: mssql
    db.name: lsnext_domain
    peer.service: srv2.local,1433
    db.statement_type: Text
Resource associated with Activity:
    service.name: some-service-name
    service.instance.id: d597a313-eeed-4e9e-a32c-79bb4a1538e8
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.7.0-alpha.1
```
We can see both expected spans using Enrich methods (pay attention to `Id` and `ParentSpanId`), so I guess that both spans are exist and looks like the first one just doesnt stop:
<img width="800" alt="image" src="https://github.com/open-telemetry/opentelemetry-dotnet/assets/32906291/ccce2699-827d-411b-ad33-a8cbe076e0b3">
<img width="800" alt="image" src="https://github.com/open-telemetry/opentelemetry-dotnet/assets/32906291/6d82b081-5fe0-4427-89f4-ad38f04c2c3d">

## Reproduce

```csharp
using Microsoft.EntityFrameworkCore;
using OpenTelemetry;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;

using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .ConfigureResource(r => r.AddService("some-service-name"))
    .AddSqlClientInstrumentation()
    .AddEntityFrameworkCoreInstrumentation()
    .AddConsoleExporter()
    .Build();

var dbContextOptions = new DbContextOptionsBuilder<SomeDbContext>()
    .UseSqlServer("Server=srv2.local,1433;Database=lsnext_domain;User Id=sa;Password=Qwerty123;trustservercertificate=true").Options;
{
    await using var dbContext = new SomeDbContext(dbContextOptions);
    await dbContext.Database.ExecuteSqlAsync($"select 1");
}

await Task.Delay(2000);
Console.ReadKey();

public class SomeDbContext : DbContext
{
    public SomeDbContext(DbContextOptions<SomeDbContext> options) : base(options)
    {
    }
}
```

## Additional Context

- Other exporters have the same behavior (tested on OpenTelemetry.Exporter.OpenTelemetryProtocol 1.6.0)

**Comment by cijothomas** [comment]
Moving to the contrib repo as EF instrumentation is hosted there? 

@virtes Do you have more idea on the underlying issue and possible fixes? There isn't a marked owner for EF instrumentation, so if you/others can help investigate and provide fix, that'd be super helpful!

**Comment by virtes** [comment]
@cijothomas yes, lets move it to the contrib repo. I will suggest potential fix there

**Comment by virtes** [comment]
@cijothomas I'm not sure, that the problem is caused by EF instrumentation.

As I understood, the underlying problem is that [EntityFrameworkDiagnosticListener](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs) and [SqlClientDiagnosticListener](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientDiagnosticListener.cs  ) are called in parallel so [stopping the SqlClient activity](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientDiagnosticListener.cs#L172) doesnt reflect [actual current activity in EF listener](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/Shared/DiagnosticSourceListener.cs#L73) and this causes the EF instrumentation to use obsolete activity. It turns out that the child SqlClient span is used [here](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs#L258), but the EF activity is expected here, so the activity doesnt stop and could not be exported

**Comment by mbakalov** [comment]
I will take a look at this one! cc @cijothomas

**Comment by markphillips100** [comment]
I've just experienced this issue.  Any progress been made?

My package versions are:
```
    <PackageVersion Include="OpenTelemetry.Exporter.Console" Version="[1.7.0, 2.0.0)" />
    <PackageVersion Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="[1.7.0, 2.0.0)" />
    <PackageVersion Include="OpenTelemetry.Extensions.Hosting" Version="[1.7.0, 2.0.0)" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.AspNetCore" Version="[1.7.0, 2.0.0)" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.EntityFrameworkCore" Version="1.0.0-beta.9" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.Http" Version="[1.7.0, 2.0.0)" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.SqlClient" Version="1.6.0-beta.3" />

```
